Refactor lambda post hook.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37025 Refactor lambda post hook.**

This allows us to reuse this framework in other places.

Differential Revision: [D20958327](https://our.internmc.facebook.com/intern/diff/D20958327/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20958327/)!

pritamdamania committed:
Refactor lambda post hook.

This allows us to reuse this framework in other places.

Differential Revision: [D20958327](https://our.internmc.facebook.com/intern/diff/D20958327/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20958327/)!

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -135,8 +117,11 @@ Reducer::Reducer(
-                [=] { this->autograd_hook(index); })),
+            grad_accumulator->add_post_hook(
+                torch::make_unique<torch::autograd::utils::LambdaPostHook>(
+                    [=](const torch::autograd::variable_list& /* unused */) {
```
what's this parameter for given that it's not used?

pritamdamania87 replied:
Its not used in this case, but you can check the internal changed in D20958327 where we use the inputs parameter. That is why we had to change the signature as well.

wanchaol replied:
I see, this make sense.

wanchaol commented on torch/csrc/autograd/utils/lambda_post_hook.h:
```diff
@@ -0,0 +1,30 @@
+  using variable_list = std::vector<torch::autograd::Variable>;
+
+ public:
+  /* implicit */ LambdaPostHook(std::function<void(const variable_list&)> fn)
```
can we document what is `const variable_list&` is for and how it's used? it's hard to know from this defintion unless seeing from it's actual usage.

albanD replied:
While you're changing this signature, why not go all the way and also allow it to modify the output. That way we are future proof and won't have to change this again?

wanchaol commented on torch/csrc/autograd/utils/lambda_post_hook.h:
```diff
@@ -0,0 +1,30 @@
+  variable_list operator()(
+      const variable_list& outputs,
+      const variable_list& inputs) override {
+    fn_(inputs);
```
I didn't read through what the lambda_post_hook is for, but is this `lambda_post_hook` only supposed to be used by AccmulateGrad? It seems that we are always passing the inputs in, although it is ok since other variables can be passed in by capture.

pritamdamania87 replied:
Currently, it is only used for AccumulateGrad.

pritamdamania committed:
Update on "Refactor lambda post hook."


This allows us to reuse this framework in other places.

Differential Revision: [D20958327](https://our.internmc.facebook.com/intern/diff/D20958327/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20958327/)!

[ghstack-poisoned]

albanD reviewed (approved):
lgtm
